module/Finna/src/Finna/Controller/ReservationListController.php

```diff
@@ -0,0 +1,471 @@
+
+/**
+ * Reservation List Controller
+ *
+ * @category VuFind
+ * @package  Controller
+ * @author   Samuli Sillanpää <samuli.sillanpaa@helsinki.fi>
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @author   Konsta Raunio <konsta.raunio@helsinki.fi>
+ * @author   Kalle Pyykkönen <kalle.pyykkonen@helsinki.fi>
+ * @author   Juha Luoma <juha.luoma@helsinki.fi>
+ * @author   Tuure Ilmarinen <tuure.ilmarinen@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     http://vufind.org   Main Site
+ */
+class ReservationListController extends AbstractBase
+{
+    use ListItemSelectionTrait;
+
+    /**
+     * Constructor
+     *
+     * @param ServiceLocatorInterface $sm                     Service locator
+     * @param ReservationListService  $reservationListService Reservation list service
+     * @param ReservationList         $reservationListHelper  Reservation list helper
+     */
+    public function __construct(
+        ServiceLocatorInterface $sm,
+        protected ReservationListService $reservationListService,
+        protected ReservationList $reservationListHelper
+    ) {
+        parent::__construct($sm);
+    }
+
+    /**
+     * Retrieves the value of the specified parameter.
+     *
+     * @param string $param   The name of the parameter to retrieve.
+     * @param mixed  $default Default value to return if not found
+     *
+     * @return mixed The value of the specified parameter.
+     */
+    protected function getParam(string $param, mixed $default = null): mixed
+    {
+        return $this->params()->fromRoute($param)
+            ?? $this->params()->fromPost($param)
+            ?? $this->params()->fromQuery($param, $default);
+    }
+
+    /**
+     * Validate CSRF from post request
+     *
+     * @return bool
+     */
+    protected function validateCsrf(): bool
+    {
+        $csrf = $this->serviceLocator->get(\VuFind\Validator\CsrfInterface::class);
+        $valueFromPost = $this->getRequest()->getPost()->get('csrf');
+        return $valueFromPost && $csrf->isValid($valueFromPost);
+    }
+
+    /**
+     * Add item to list action.
+     *
+     * @return \Laminas\View\Model\ViewModel
+     * @throws \Exception
+     */
+    public function addItemToListAction()
+    {
+        $user = $this->getUser();
+        if (!$user) {
+            return $this->forceLogin();
+        }
+        $view = $this->createViewModel(
+            [
+                'institution' => $this->getParam('institution'),
+                'listIdentifier' => $this->getParam('listIdentifier'),
+                'recordId' => $this->getParam('recordId'),
+                'source' => $this->getParam('source'),
+            ]
+        );
+        $driver = $this->getRecordLoader()->load(
+            $view->recordId,
+            $view->source ?: DEFAULT_SEARCH_BACKEND,
+            false
+        );
+        [
+            'properties' => $listProperties,
```

Comment:
Mä en ole koskaan ennen nähnyt tätä syntaksia. Ovelaa. Mutta useimmiten tarpeeton ja hämääväkin. Nyt tehdään muuttujia, joita ei tarvita. Jos tarvitaan vain yhtä elementtiä, kuten tässä, olisi selkeämpää ottaa vain se:
`$listProperties = ($this->reservationListHelper)($user)->getListProperties($view->institution, $view->listIdentifier)['properties'];`
Muissa kohdissa jättäisin elementit assosiatiiviseen taulukkoon, ellei ole painavaa syytä ottaa niitä erilleen.